src/src/com/microsoft/aad/adal/HttpWebRequest.java

```diff
@@ -107,107 +92,116 @@ private void setupConnection() {
             throw new IllegalArgumentException("requestURL");
         }
         HttpURLConnection.setFollowRedirects(true);
-        mConnection = openConnection();
+        final HttpURLConnection connection = (HttpURLConnection)mUrl.openConnection();
+        connection.setConnectTimeout(CONNECT_TIME_OUT);
         // To prevent EOF exception.
         if (Build.VERSION.SDK_INT > 13) {
-            mConnection.setRequestProperty("Connection", "close");
+            connection.setRequestProperty("Connection", "close");
         }
+
+        // Apply the request headers
+        final Iterator<String> headerKeys = mRequestHeaders.keySet().iterator();
+
+        while (headerKeys.hasNext()) {
+            String header = headerKeys.next();
+            Logger.v(TAG, "Setting header: " + header);
+            connection.setRequestProperty(header, mRequestHeaders.get(header));
+        }
+
+        // Avoid reuse of existing sockets to avoid random EOF errors
+        System.setProperty("http.keepAlive", "false");
+        connection.setReadTimeout(READ_TIME_OUT);
+        connection.setInstanceFollowRedirects(mInstanceRedirectsFollow);
+        connection.setUseCaches(mUseCaches);
+        connection.setRequestMethod(mRequestMethod);
+        connection.setDoInput(true); // it will at least read status
+                                     // code. Default is true.
+        setRequestBody(connection, mRequestContent, mRequestContentType);
+
+        return connection;
     }
 
     /**
      * send the request.
-     * 
-     * @param contentType
      */
-    public HttpWebResponse send() {
-
-        Logger.v(TAG, "HttpWebRequest send thread:" + android.os.Process.myTid());
-        setupConnection();
-        HttpWebResponse response = new HttpWebResponse();
-
-        if (mConnection != null) {
+    public HttpWebResponse send() throws IOException {
+        Logger.v(TAG, "HttpWebRequest send thread:" + Process.myTid());
+        final HttpURLConnection connection = setupConnection();
+        final HttpWebResponse response;
```

Comment:
it would compile - it works because variable initialized only once and in all code paths